crates/uv-interpreter/src/discovery.rs

```diff
@@ -0,0 +1,1134 @@
+                    toolchains.filter(move |toolchain|
+                        version.is_none() || version.is_some_and(|version|
+                            version.matches_version(toolchain.python_version())
+                        )
+                    )
+                    .map(|toolchain| (InterpreterSource::ManagedToolchain, toolchain.executable()))
+                )
+                .map_err(Error::from)
+            ).flatten_ok()
+        ).into_iter().flatten()
+    )
+    // (4) The search path
+    .chain(
+        sources.contains(InterpreterSource::SearchPath).then(move ||
+            python_executables_from_search_path(version)
+            .map(|path| Ok((InterpreterSource::SearchPath, path))),
+        ).into_iter().flatten()
+    )
+    // (5) The `py` launcher (windows only)
+    // TODO(konstin): Implement <https://peps.python.org/pep-0514/> to read python installations from the registry instead.
+    .chain(
+        (sources.contains(InterpreterSource::PyLauncher) && cfg!(windows)).then(||
+            std::iter::once(
+                py_list_paths()
+                .map(|entries|
+                    // We can avoid querying the interpreter using versions from the py launcher output unless a patch is requested
+                    entries.into_iter().filter(move |entry|
+                        version.is_none() || version.is_some_and(|version|
+                            version.has_patch() || version.matches_major_minor(entry.major, entry.minor)
+                        )
+                    )
+                    .map(|entry| (InterpreterSource::PyLauncher, entry.executable_path))
+                )
+                .map_err(Error::from)
+            ).flatten_ok()
+        ).into_iter().flatten()
+    )
+}
+
+/// Lazily iterate over Python executables in the `PATH`.
+///
+/// The [`VersionRequest`] is used to determine the possible Python interpreter names, e.g.
+/// if looking for Python 3.9 we will look for `python3.9` in addition to the default names.
+///
+/// Executables are returned in the search path order, then by specificity of the name, e.g.
+/// `python3.9` is preferred over `python3`.
+///
+/// If a `version` is not provided, we will only look for default executable names e.g.
+/// `python3` and `python` — `python3.9` and similar will not be included.
+fn python_executables_from_search_path(
+    version: Option<&VersionRequest>,
+) -> impl Iterator<Item = PathBuf> + '_ {
+    // `UV_TEST_PYTHON_PATH` can be used to override `PATH` to limit Python executable availability in the test suite
+    let search_path =
+        env::var_os("UV_TEST_PYTHON_PATH").unwrap_or(env::var_os("PATH").unwrap_or_default());
+    let possible_names = version.unwrap_or(&VersionRequest::Default).possible_names();
+
+    // Split and iterate over the paths instead of using `which_all` so we can
+    // check multiple names per directory while respecting the search path order
```

Comment:
It seems like `which_in_global` does accept multiple directories, but I'm guessing this gives us slightly different ordering / preference resolution?

Reply:
Yeah we need to check each directory separately since we are checking for multiple names per directory, whereas `which_in_global` would resolve a single name.